lxd/network/network_utils.go

```diff
@@ -639,6 +639,109 @@ func randomSubnetV6() (string, error) {
 	return "", fmt.Errorf("Failed to automatically find an unused IPv6 subnet, manual configuration required")
 }
 
+// randomAddressInSubnet finds a random address within a given subnet. If given, the validate function is called for
+// each randomly generated value. Each time validate returns an error, another IP address will be generated until the
+// context is cancelled.
+func randomAddressInSubnet(ctx context.Context, ipnet *net.IPNet, validate func(net.IP) error) (net.IP, error) {
+	var bufferLen int
+	var ipVersion int
+	if ipnet.IP.To4() != nil {
+		ipVersion = 4
+		bufferLen = 4
+	} else {
+		ipVersion = 6
+		bufferLen = 16
+	}
+
+	ones, _ := ipnet.Mask.Size()
+
+	// The number of bytes that will be the same in the subnet and in the random address.
+	quotient := ones / 8
+
+	// The remaining bits that will be the same in the subnet and in the random address (after the equivalent bytes).
+	remainder := ones % 8
+
+	// Can't get a random address from e.g. 192.0.2.1/32 as it's a single IP.
+	if quotient == bufferLen {
+		return nil, fmt.Errorf("Cannot find random address when subnet contains only one address")
+	}
+
+	for {
+		// Exit if the context is cancelled.
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		// Read randomness into a byte slice of the correct length for the IP version.
+		r := make([]byte, bufferLen)
+		_, err := cryptorand.Read(r)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to automatically find an unused IPv%d address within %q: %w", ipVersion, ipnet.String(), err)
+		}
+
+		// Overwrite bytes/bits from the given subnet where the random address should be prefixed.
+		for i := 0; i <= quotient; i++ {
+			// For IPv4 addresses, the IP bytes SHOULD BE stored in the last four bytes of a 16 byte slice.
```

Comment:
You should use ip.To4() to ensure you are using the 4 byte representation always.

> To4 converts the IPv4 address ip to a 4-byte representation.

Reply:
either here or earlier in the function so we know what we are dealing with.

Reply:
This is what dnsmasq.DHCPAllAllocations does in the LXD code base btw.

Reply:
This also allows storing the IPs as keys in map.